File: Solution1.java
// Leetcode problem : https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/description/
// Time Complexity : O(logn)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach in three sentences only
/*
 * I perform binary search to find first occurrence of the target element. If it's not present, I straight away return {-1,-1}. If it's present, I store the first occurrence in ans[0] and I run a right
 * binary search from the boundaries ans[0] to nums.length-1.
 */
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int []ans = {-1,-1};
        int leftResult = lsearch(nums, target);
        if(leftResult == -1)
        {
            return ans;
        }
        ans[0]=leftResult;
        ans[1] = rsearch(nums,target,leftResult);
        return ans;
    }

    public int lsearch(int []nums, int target)
    {
        int low = 0, high = nums.length-1;
        int result = -1;
        while(low<=high)
        {
            int mid = low + (high-low)/2;
            if(nums[mid]>=target)
            {
                if(nums[mid]==target)
                    result = mid;
                high=mid-1;
            }
            else
            {
                low=mid+1;
            }
        }
        return result;
    }

    public int rsearch(int []nums, int target, int low)
    {
        int high = nums.length-1;
        int result = low;
        while(low<=high)
        {
            int mid = low + (high-low)/2;
            if(nums[mid]==target)
            {
                result = mid;
                low = mid+1;
            }
            else if(nums[mid]>target)
            {
                high = mid-1;
            }
        }
        return result;

    }
}

File: solution1.py
# Time Complexity : O(log(n))
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach

class Solution:
    def search_range(self, nums, target):
        if len(nums) == 0: return [-1, -1]
        
        # search for first position 
        def search_low(nums, target):
            head, tail = 0, len(nums) - 1
            while head <= tail:
                mid = (head + tail)//2
                if nums[mid] >= target:
                    tail = mid - 1
                else:
                    head = mid + 1
            return head
        # search for seconf position       
        def search_high(nums, target):
            head, tail = 0, len(nums) - 1
            while head <= tail:
                mid = (head + tail)//2
                if nums[mid] > target:
                    tail = mid - 1
                else:
                    head = mid + 1
            return tail
        
        start = search_low(nums, target)
        end = search_high(nums, target)
        if start <= end:
            return [start, end]
        else:
            # does not exist 
            return [-1, -1]